[1단계 - 점심 뭐 먹지] 기린(정유정) 미션 제출합니다

jeongyou opened the pull request:
안녕하세요 하리~! 기린이라고 합니다. 반갑습니다! 리뷰 잘 부탁드립니다~~🙏

## 학습 목표

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

- 어플리케이션을 컴포넌트 단위로 모듈화하여 개발
  - UI를 컴포넌트 단위로 생각하고 개발하는 연습
  - 재사용할 수 있는 컴포넌트를 고민해보기
- 웹 UI 환경에서의 테스트 기초
  - 사용자 관점에서 중요하다고 생각하는 기능을 스스로 정의하고 E2E 테스트로 검증해보기

## 제출 전 체크 리스트

<!-- 리뷰를 요청하기 전, 다음 항목들을 기본적으로 확인해 주세요. -->

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항을 준수하고 있는지 확인했나요?
- [x] 테스트 코드는 모두 정상적으로 실행되나요?
- [x] (해당하는 경우) 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포한 링크 기입: https://jeongyou.github.io/javascript-lunch/

## 리뷰 요청 & 논의하고 싶은 내용

<!-- PR 작성자로서, 코드만으로는 알기 어려운 작성자의 의도와 문제 해결 과정에 대해 공유해 주세요.
이 PR에서 달성해야 하는 학습 목표를 잘 달성하고 있는지 스스로 확인하며, 피드백 받을 수 있는 내용일지 점검해 보세요.
리뷰어에게 정답을 묻기보다 고민하고 의사 결정한 과정에 대해 공유하고 이에 대한 피드백을 받으며 대화해보기를 권장합니다. -->

### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

<!-- 구현 과정에서 가장 어려웠던 점이나 많이 고민한 점은 무엇인가요?
이를 해결하기 위해 어떤 방법들을 검토하고 시도했으며, 그 과정에서 새롭게 배운 점이 있나요? -->
재사용 가능한 컴포넌트를 만들기 위해 노력했습니다. 특히, 컴포넌트의 단위를 어떻게 나눌지 판단하는 것이 가장 어려운 부분이었습니다.
너무 세밀하게 쪼개면 오히려 재사용성이 떨어질 것 같았고, 그렇다고 너무 크게 묶으면 유연성이 부족해질 것 같아 적절한 기준을 찾는 데 많은 고민이 필요했습니다.

페어와 논의한 결과, 카테고리, 이름, 거리, 설명, 링크와 같이 반복되는 구조가 있다는 점을 발견했고, 이를 하나의 컴포넌트로 묶기로 했습니다.
또한, 모달 내부에서도 반복되는 패턴이 존재한다는 점을 인식하고, 이를 컴포넌트화하여 코드의 중복을 줄일 수 있도록 리팩토링했습니다.

이 과정에서 컴포넌트의 재사용성이 주는 장점을 직접 체감할 수 있었습니다.
데이터만 변경하면 동일한 컴포넌트를 여러 번 사용할 수 있어 코드가 간결해지고 유지보수도 쉬워졌습니다.
이러한 경험을 통해 효율적인 컴포넌트 분리와 재사용의 중요성을 다시 한번 깨닫게 되었습니다.

### 2) 이번 리뷰를 통해 논의하고 싶은 부분

<!-- 구현한 코드와 학습 목표와 관련해 피드백을 받고 싶은 부분이나, 함께 논의해보고 싶은 점 -->
**1. 파일의 구조**
- 파일의 구조를 아래와 같이 구성해보았는데 하리 생각에는 잘 분리가 된 것인지 궁금합니다.
```
📦 src                        # 프로젝트 소스 코드 폴더
 ┣ 📂 components              # 재사용 가능한 UI 컴포넌트 모음
 ┃ ┣ 📜 Button.js            # <button> 요소를 생성하는 컴포넌트
 ┃ ┣ 📜 Header.js            # <header> 요소를 생성하는 컴포넌트 (앱 제목 표시)
 ┃ ┣ 📜 index.js             # components 폴더 내 모든 컴포넌트를 한 곳에서 export
 ┃ ┣ 📜 Input.js             # <input> 요소를 생성하는 컴포넌트 (텍스트 입력 필드)
 ┃ ┣ 📜 Modal.js             # <div class="modal"> 요소를 생성하는 컴포넌트 (모달 UI)
 ┃ ┣ 📜 RestaurantItem.js    # <li> 요소를 생성하는 컴포넌트 (음식점 리스트 항목)
 ┃ ┣ 📜 Select.js            # <select> 요소를 생성하는 컴포넌트 (드롭다운 메뉴)
 ┃ ┗ 📜 Textarea.js          # <textarea> 요소를 생성하는 컴포넌트 (설명 입력 필드)
 ┣ 📜 constants.js           # 변하지 않는 상수 데이터 저장 (카테고리, 거리, 버튼 텍스트 등)
 ┣ 📜 eventHandlers.js       # 사용자 입력 이벤트 핸들러 관리 (모달 열기, 음식점 추가 등)
 ┣ 📜 main.js                # 애플리케이션 진입점 (UI 생성 및 이벤트 핸들러 등록)
 ┣ 📜 restaurantData.js      # 초기 음식점 데이터를 저장하는 파일
 ┣ 📜 stateStore.js          # 애플리케이션 상태 관리 (새 음식점 데이터 저장 및 관리)
 ┣ 📜 style.css              # 애플리케이션의 스타일 정의 (UI 디자인 및 레이아웃)
 📜 index.html             # 웹 애플리케이션의 기본 HTML 구조
```

**2. required 적용 방식에 대한 고민**
-  constants.js 파일을 보시면 
```
export const NAME = {
  label: '이름',
  name: 'name',
  helpText: '',
  required: 'form-item--required',
  type: 'text',
};

export const LINK = {
  label: '참고 링크',
  name: 'link',
  helpText: '매장 정보를 확인할 수 있는 링크를 입력해 주세요.',
  required: '',
  type: 'url',
};
```
- 이런식으로 required 정보까지 저장해서 컴포넌트에 전달하여 required를 지정해주고 있습니다. 이 방법이 적절한 것인지 아니면 다른 방법이 있는지 궁금합니다.

**3. createElement 방식과 현재 방식(템플릿 리터럴 방식) 중 어떤 것이 더 적절한지에 대한 질문**
- 현재 저희가 작성한 컴포넌트들은 템플릿 리터럴을 사용하여 HTML 문자열을 반환하는 방식으로 구현했습니다.
이 방식의 장점은 간결한 코드와 직관적인 HTML 구조 유지라고 생각합니다. 하지만, 직접 DOM 요소를 생성하는 createElement 방식이 더 적절할 수도 있다고 생각이 들기도 하는데 하리는 어떤 방식을 더 선호하시나요?

**4. 컴포넌트 분리에 대한 질문**
저희는 '반복되는 로직'을 기준으로 컴포넌트를 나누었고, 먼저 Templates 파일을 기반으로 공통된 패턴을 정리한 후 컴포넌트로 분리했습니다.
이 방식이 재사용성과 유지보수 측면에서 적절한지, 혹시 더 효과적인 분리 기준이 있을지 하리의 의견이 궁금합니다!

**5. E2E 테스트**
- 현재 E2E 테스트에서는 사용자의 주요 행동 시나리오를 기반으로 테스트를 작성했습니다. 현재 작성된 테스트 시나리오가 적절한 범위를 커버하고 있는지 하리의 의견이 궁금합니다! 혹시 놓치고 있는 중요한 시나리오가 있을까요?

**6. e2e 테스트에서 `{ force: true }` 사용**
e2e 테스트 코드를 작성하던 중, cypress가 modal을 인식하지 못하는 문제로 인해, 다음과 같이 코드에 `{ force: true }` 속성을 부여했습니다.
이렇게 요소가 안 보이더라도 강제 클릭하게 하는 기능을 사용하여, 테스트를 통과하는 것이 적절한 방식일지 고민입니다.
하리의 의견은 어떠신가요?
```javascript
 it('사용자가 음식점 추가 모달을 연 후, ESC 키를 클릭하여 음식점 추가를 취소하는 시나리오 테스트', () => {
    cy.get('.gnb__button').click();
    cy.get('.modal--open').should('exist');
    cy.document().trigger('keydown', { key: 'Escape', keyCode: 27, which: 27, force: true });
    cy.get('.modal--open').should('not.exist');
  });
```
---

<details>
<summary>✅ 리뷰어 체크 포인트</summary>
<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.-->

### 1단계

- [ ] UI를 컴포넌트 단위로 분리했는가?
- [ ] 재사용 가능한 컴포넌트를 고려했는가?
- [ ] 주요 기능 시나리오에 대한 E2E 테스트를 작성했는가?

### 2단계

- [ ] 오류 상황(예: 잘못된 입력)에 대해서도 테스트 케이스를 정의했는가?
- [ ] 타입 정의를 적절히 사용하여 컴파일 오류 없이 안정적으로 동작하는가?
- [ ] 로컬/배포 환경에서 테스트가 정상적으로 동작하는지 확인했는가?

</details>

JeLee-river committed:
chore: eslint, prettier 설정

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
docs: 기능 목록 정리

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
chore: 이미지 파일을 public으로 이동

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
style: 전체 레이아웃 및 스타일 적용

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
feat: 동적으로 GNB 헤더 생성 및 페이지 로드시 추가

- createHeader 함수 추가하여 헤더 동적 생성
- window load 이벤트를 사용해 페이지 로드시 헤더 자동 삽입
- 기존 정적인 HTML 대신 JavaScript로 GNB 구조 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
style: 전체 레이아웃 중앙 정렬

- body와 html에 flexbox를 적용하여 전체 페이지 중앙 정렬

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
docs: 헤더 생성 구현 완료에 따른 구현할 기능 목록 업데이트

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
feat: 동적으로 음식점 아이템 리스트 생성 및 페이지 로드 시 추가

- 페이지 로드 시 음식점 리스트를 동적으로 생성하여 `.restaurant-list`에 추가
- `RESTAURANTS` 데이터를 기반으로 `createRestaurantItem` 함수를 활용하여 리스트 아이템을 생성
- `insertAdjacentHTML('beforeend', items)`를 사용하여 기존 리스트에 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
docs: 테스트 목록 작성

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
refactor: 상수를 constants.js 파일로 분리

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
test: app의 주요 ui 요소 렌더링 테스트 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
feat: 음식점 추가 모달을 여는 기능 구현

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
feat: 새 음식점 정보를 받아와서 저장하는 기능 구현

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
feat: 모달의 추가하기 버튼을 누르면 음식점 아이템을 추가하는 기능 구현

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
feat: 취소하기 버튼을 누르면 모달을 닫는 기능 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
refactor: main.js에서 lunchUI.js로 ui 로직 분리

- main.js는 애플리케이션 로직을 담당하도록 유지
- UI 관련 함수는 lunchUI.js에서 관리

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
feat: 취소하기 버튼, 추가하기 버튼을 눌렀을 때 입력을 초기화 하고 닫는 기능 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
style: 모달 너비 조정

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
feat: 모달 외부를 클릭했을때 입력을 초기화하고 닫는 기능 구현

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
fix: enter, space 키를 눌러도 모달이 닫히는 현상 수정

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
docs: 구현 완료한 기능 체크

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
feat: 모달창 UI 구현

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
test: 모달 렌더링 테스트 추가

JeLee-river committed:
test: constants.js 파일 변경에 따른 테스트 로직 수정

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
docs: e2e 테스트 목록 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
test: 음식점 추가 모달을 열어, 모든 필드를 입력 후 음식점을 추가하는 시나리오 테스트 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
test: 음식점 추가 모달을 열어, 필수 입력값만 입력 후 추가하기 시나리오 테스트 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
test: 음식점 추가 모달을 연 후, 음식점 추가 중 취소하기 버튼을 클릭하여 음식점 추가를 취소하는 시나리오 테스트 추가

JeLee-river committed:
test: modal이 열리는지 확인하는 로직 수정

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
test: 음식점 추가 모달을 연 후, ESC 키를 클릭하여 음식점 추가를 취소하는 시나리오 테스트 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
test: 음식점 추가 모달을 연 후, 모달 외부를 클릭하여 음식점 추가를 취소하는 시나리오 테스트 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
test: 음식점 추가 모달을 연 후, 모달을 닫았다가 다시 열었을 때 필드가 초기화 되는 시나리오 테스트 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
docs: 모달이 렌더링 되는지 확인하는 렌더링 테스트 체크

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
feat: input link의 type을 url로 설정

- 그에 따른 테스트 코드 변경
- constants.js에 type 속성 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
test: 모달의 필수 필드 입력창에 아무것도 입력하지 않고 추가하기 버튼을 클릭하는 시나리오 테스트 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
refactor: lunchUI를 컴포넌트별로 분리

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
refactor: 상수 파일에서 레스토랑 데이터를 restaurantData.js로 분리

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
refactor: 거리 데이터를 숫자형으로 변경

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
refactor: UI 생성 로직을 함수로 분리하여 코드 가독성 개선

- 헤더 추가 로직을 appendHeader()로 분리
- 음식점 리스트 초기화 로직을 initRestaurantItems()로 분리
- 모달 생성 로직을 appendModal()로 분리
- 모달 내부 컨텐츠 추가 로직을 appendModalContents()로 분리
- 버튼 추가 로직을 appendModalButton()으로 분리
- 이벤트 핸들러 등록 로직을 addEventHandlers()로 분리

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
refactor: 버튼 추가 로직을 추가로 분리

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

jeongyou committed:
refactor: 이미지 파일의 절대 경로 수정

jeongyou committed:
chore: 배포 주소 설정

LAH1203 commented on cypress/e2e/featureTest.cy.js:
```diff
@@ -0,0 +1,87 @@
+
+    cy.get('.button--primary').click();
+    cy.get('.restaurant-list').should('contain', '꺼벙이');
+  });
```
앞의 전제가 겹치는 테스트가 몇 가지 있는 것 같아, DCI 패턴을 사용해 테스트에 deps를 주면 터미널에서 테스트의 가독성이 좀 더 좋아질 것 같아요~!

jeongyou replied:
DCI 패턴이라는 개념이 있었군요!
저도 테스트를 단순히 나열하기보다 계층화를 시키는 것이 좋다고 판단하여, `describe()` 블록을 활용해 테스트를 목적별로 나눠보았습니다.  또, 리팩터링 중 모달을 여는 동작이 반복된다는 점을 발견해서 `beforeEach()`에서 한 번만 실행되도록 개선했습니다.  

관련 커밋: [test: DCI 패턴을 적용하여 Cypress 테스트 계층화 및 중복 코드 제거](https://github.com/woowacourse/javascript-lunch/pull/216/commits/93fc1672bbbfa6776ec86c66201ed254a4682894)

LAH1203 commented on index.html:
```diff
@@ -3,11 +3,21 @@
-    <div id="app"></div>
-    <script type="module" src="/src/main.js"></script>
+    <div id="app">
+      <script type="module" src="./src/main.js"></script>
```
script 태그를 div 안에 옮겨두신 이유가 있을까요?

jeongyou replied:
앗! 실수로 `<script>` 태그를 `<div id="app">` 내부에 넣었습니다.
보통 `<script>`는 `<body>` 태그의 하단에 배치하여, HTML이 먼저 렌더링된 후 JavaScript가 실행되도록 하기 때문에 해당 `<script>` 태그를 `<div id="app">` 바깥으로 옮기는 것이 더 적절하다고 판단하여 수정했습니다!  

관련 커밋: [refactor: script 태그 위치 조정하여 HTML 구조 개선](https://github.com/woowacourse/javascript-lunch/pull/216/commits/705b5f8a886352eb4f5905a55e34d86538e0ebe3)

LAH1203 replied:
참고 : 기존대로 유지하고 script 태그에 defer나 async attr를 추가해도 좋아요~!

LAH1203 commented on src/components/Modal.js:
```diff
@@ -0,0 +1,13 @@
+    </div>`;
+
+  return modal;
+}
```
정적인 템플릿 반환을 함수로 구현해두신 이유도 궁금해요! [template 태그](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/template)를 적용해봐도 좋을 것 같아서요 :-)

jeongyou replied:
처음에는 별다른 고민 없이 다른 컴포넌트들처럼 모달을 동적으로 생성하는 방식이 자연스럽다고 생각해서 함수로 구현했습니다.  
그런데 피드백을 받고, 정적인 템플릿을 함수로 만들기보다는 그냥 HTML에 직접 작성하는 것이 더 적절할 수도 있겠다고 생각하였습니다.  

현재 미션에서는 모달이 한 번만 사용되기 때문에 `<template>` 태그를 사용하지 않고 단순히 HTML에 작성할까 고민했지만, 추후에 모달을 동적으로 여러 개 추가해야 할 가능성도 고려하여 `<template>` 태그를 활용하는 것이 더 나을 것 같아 적용해보았습니다!  

관련 커밋: [refactor: 모달 생성 방식을 template 태그 기반으로 변경](https://github.com/woowacourse/javascript-lunch/pull/216/commits/c4326ecc4b37f4e48318eab718ea78486d57a5e5)

이 피드백을 통해 `<template>` 태그를 처음 알게 되었는데, 컴포넌트는 데이터의 변화에 따라 UI가 동적으로 렌더링될 때 유용하고, `<template>` 태그는 초기에는 렌더링되지 않지만, 동일한 HTML 구조를 JavaScript에서 여러 번 활용할 때 좋을 것 같다는 생각이 들었습니다!

LAH1203 commented on src/components/Select.js:
```diff
@@ -0,0 +1,15 @@
+function createSelect(fieldName) {
+  const select = `<div class="form-item form-item--required">
+    <label for="${fieldName.name} text-caption">${fieldName.label}</label>
```
label for은 select에 연결해주기 위해서 추가해주신 걸까요...? 근데 지금 내용이 일치하지는 않아서 별도로 연결된 형태가 아닌 것 같아요!

jeongyou replied:
label의 `for` 속성이 올바르게 설정되지 않았던 걸 이제 확인했네요!  
select 뿐만 아니라 input, textarea 도 `label`과 `id` 값이 일치하지 않아서  실제로 `label`을 클릭해도 해당 필드가 활성화되지 않는 문제가 있었습니다.  

불필요하게 포함된 `"text-caption"` 문자열을 제거하고, 스타일을 유지하기 위해 `class="text-caption"`을 별도로 추가하는 방식으로 수정했습니다.!

관련 커밋: [fix: label 태그의 for 속성을 올바르게 수정하여 입력 필드와 연결](https://github.com/woowacourse/javascript-lunch/pull/216/commits/3398a6d20027cd6b2f2a3f4afd30fbdf45bbf228)

LAH1203 commented on src/eventHandlers.js:
```diff
@@ -0,0 +1,80 @@
+    const modal = document.querySelector('.modal');
+
+    modal.classList.add('modal--open');
+  });
```
전체적으로 이벤트 핸들러 등록에 익명 함수를 사용해주셨네요! 이렇게 했을 때 우려되는 점은 어떤 게 있을까요?

jeongyou replied:
관련 내용을 찾아보면서 익명 함수를 이벤트 핸들러로 등록할 때 발생할 수 있는 문제를 새롭게 알게되었습니다. 같은 이벤트를 여러 요소에서 사용할 경우 **매번 새로운 함수를 생성해야 해서 재사용이 어렵다는 점**과,  
이벤트를 해제해야 할 경우 **익명 함수는 제거할 수 없다는 점**입니다.

그래서 이벤트 핸들러를 개별 함수로 분리하여 재사용 가능하도록 개선하고, 제거가 필요할 경우에는`removeEventListener()`를 사용할 수 있도록 변경해보았습니다!

관련 커밋: [refactor: 이벤트 핸들러 구조 개선 및 main.js에 적용](https://github.com/woowacourse/javascript-lunch/pull/216/commits/5bf1511f0c3bbd4df6f700f8993b0ce859d0c8fe)

LAH1203 commented on src/eventHandlers.js:
```diff
@@ -0,0 +1,80 @@
+  modalBackdrop.addEventListener('click', () => {
+    const modal = document.querySelector('.modal');
+
+    resetFormAndState();
```
모달이 닫힐 때마다 데이터를 초기화시켜주셔서 넘 편해요~!~!

LAH1203 commented on src/main.js:
```diff
@@ -1,23 +1,94 @@
+  const nameInputElement = document.querySelector('#name');
+  setRequired(nameInputElement);
+
+  addEventHandlers();
```
이벤트 핸들러의 해제는 신경쓰지 않아도 괜찮을까요?

jeongyou replied:
현재 코드에서는 `registerEventHandlers()`가 `DOMContentLoaded`에서 한 번만 실행되기 때문에, 이벤트 핸들러가 중복 등록될 가능성이 없어 이벤트 해제를 꼭 할 필요는 없다고 생각했습니다.  

그리고 현재 UI 구조에서는 모달이 완전히 삭제되는 것이 아니라 단순히 숨겨지는 형태이므로, 이벤트 핸들러가 남아 있어도 큰 문제가 되지 않는다고 판단했습니다.  

하지만 나중에 UI가 동적으로 변경될 경우 이벤트 해제가 필요할 수도 있다고 생각하여 `removeEventHandlers()`를 추가해 두었습니다.  

관련 커밋: [refactor: 이벤트 핸들러 구조 개선 및 main.js에 적용](https://github.com/woowacourse/javascript-lunch/pull/216/commits/5bf1511f0c3bbd4df6f700f8993b0ce859d0c8fe)

저는 현재는 이벤트 해제가 필수적이지 않다고 판단했지만, 혹시 하리는 현재 코드에서도 이벤트 해제가 필요하다고 보시는지 궁금합니다..! 제가 놓친 부분이 있다면 피드백 주시면 감사하겠습니다!

LAH1203 replied:
지금처럼 모달이 초기 돔에서 같이 렌더링 되는 형태에서는 크게 필요하지 않을 것 같습니다~
하지만 말씀해주신대로 모달이 돔에서 동적으로 추가 / 해제되는 동작이 반복된다면, 등록된 이벤트 핸들러를 해제하지 않아도 정상적으로 동작할 수는 있지만, 해제하지 않은 이벤트 핸들러가 메모리에 남아 있는 경우도 배제할 수 없을 것 같아요! 그래서 이벤트 해제를 습관화하시면 좋을 것 같아서 말씀 드렸습니다 :-)
관련해서 하면 좋다, 안해도 된다 의견이 많던데, 그래서 몇 개 토론 글을 가져와봤어요~! 심심하실 때 한번씩 보셔요 !
https://www.reddit.com/r/Frontend/comments/1ato11w/will_the_event_listeners_be_removed_automatically/?rdt=42559
https://forum.freecodecamp.org/t/remove-events-for-garbage-collector-to-collect/439517
https://stackoverflow.com/questions/12528049/if-a-dom-element-is-removed-are-its-listeners-also-removed-from-memory

LAH1203 commented on src/main.js:
```diff
@@ -1,23 +1,94 @@
+  const main = document.querySelector('main');
+  const modal = createModal();
+
+  main.insertAdjacentHTML('beforeend', modal);
```
아까 다른 곳에서는 innerHTML을 사용해주셨는데, 여기에서는 insertAdjacentHTML를 사용해주신 이유가 궁금해요~!

jeongyou replied:
components/Header.js 파일에서 사용한 innerHTML 말씀하시는 것 맞나요??!

저는 헤더는 사용자 입력과 관련 없는 정적인 요소라 `innerHTML`을 사용해도 안전하다고 생각했고, 사용자 입력이 포함될 가능성이 있는 요소에서는 XSS 공격을 방지하기 위해  
`insertAdjacentHTML()`을 사용해 보았습니다!  

❓한가지 궁금한 점은 통일성을 위해 헤더에서도 `insertAdjacentHTML()`을 사용하는 것이 나을까요? 어떤 방식이 더 좋을지 하리의 의견이 궁금합니다!

LAH1203 replied:
네네 헤더 쪽에서 쓰인 `innerHTML`을 말씀드린 게 맞습니다~!
XSS 공격을 방지하기 위해 `insertAdjacentHTML`을 사용해주셨다고 했는데, 사용자 입력 값을 그대로 사용하는 건 결국 `innerHTML`나 `insertAdjacentHTML` 모두 동일하게 XSS 위험이 있다고 생각합니다 🥹
딱히 통일을 바라고 언급했던 건 아니었고, 둘의 차이를 알고 계신지, 알고 계시다면 왜 따로 나눠서 적용해주신 건지 궁금해서 여쭤봤습니다 ㅎㅎ XSS 공격 방지를 위해서라면 지금과 다른 방식을 고민해보셔도 좋을 것 같아요!

LAH1203 commented on src/stateStore.js:
```diff
@@ -0,0 +1,29 @@
+const stateStore = {
```
스토어 설계라니... 👍👍

jeongyou replied:
페어가 제안해 주어서 적용해 보았는데 잘 한 것이었군요!!

LAH1203 replied:
flux 패턴이 생각나는 설계였어요~! 😄
자바스크립트로 직접 구현해주셔서, 좋은 시도를 해주신 것 같습니당ㅎㅎ

LAH1203 reviewed (changes requested):
안녕하세요 기린 🦒 만나서 반가워요~!~!

전체적으로 깔끔하게 잘 분리해주셨네요! 컴포넌트 재사용성을 고려해서 작성해주신 게 인상 깊었습니당
사용성을 고려해서 모바일 기준 UI로 만드시고, 모달 (바텀싯) 을 닫을 때마다 데이터를 초기화해주신 점 등 여러모로 UX를 잘 고려해주신 것 같아요~!

기린이 PR에서 질문 주신 부분은 마지막 코멘트로 남겨두겠습니다 :-)
지금 머지해도 무리가 없을 것 같은데, 코멘트 확인 후에 수정하고 싶으신 부분이 생길까봐 일단 request changes로 드립니당. 수정하지 않아도 괜찮다고 생각하시면 바로 리뷰요청 주세요~!

LAH1203 commented:
> 파일의 구조를 아래와 같이 구성해보았는데 하리 생각에는 잘 분리가 된 것인지 궁금합니다.

분리 자체는 잘 해주신 것 같아요! 다만 지금은 src 내에서 컴포넌트만 따로 폴더링을 해주셨는데, 다른 것들(constants, store, styles 등)도 따로 폴더링이 되면 일관성 측면에서 더 좋을 것 같아요~

> 이런식으로 required 정보까지 저장해서 컴포넌트에 전달하여 required를 지정해주고 있습니다. 이 방법이 적절한 것인지 아니면 다른 방법이 있는지 궁금합니다.

required는 지금처럼 각 attr 객체에서 가지고 있어도 괜찮다고 생각해요! 근데 보통 required는 boolean으로 예상할텐데, 문자열이어서 그건 조금 낯설었어요 😅

> 현재 저희가 작성한 컴포넌트들은 템플릿 리터럴을 사용하여 HTML 문자열을 반환하는 방식으로 구현했습니다.  이 방식의 장점은 간결한 코드와 직관적인 HTML 구조 유지라고 생각합니다. 하지만, 직접 DOM 요소를 생성하는 createElement 방식이 더 적절할 수도 있다고 생각이 들기도 하는데 하리는 어떤 방식을 더 선호하시나요?

저는 createElement를 더 선호합니당. 코드를 어떻게 작성하느냐에 따라서 createElement를 사용하더라도 충분히 가독성 있게 작성할 수 있기도 하고 무엇보다 XSS 공격을 방지할 수 있다는 점이 가장 큰 장점이라고 생각해요.
템플릿 리터럴 방식은 간결하고 직관적이지만, 사용자 입력을 포함해야 하는 경우 innerHTML을 사용하면 스크립트 삽입 공격에 취약해질 가능성이 있습니당. 반면에 createElement를 사용하면 DOM 요소를 직접 생성하고 조작하기 때문에 안전성이 더 높아지고, 브라우저가 HTML을 파싱하는 과정도 줄어들어 성능 면에서도 유리할 수 있어요!
또, 요소를 동적으로 추가해야 할 때 createElement를 사용하면 이벤트 핸들러를 더욱 직관적이고 효율적으로 관리할 수 있어요. 템플릿 리터럴 방식에서는 사용해주신 것처럼 innerHTML나 insertAdjacentHTML를 사용하게 되는데, 이 과정에서 기존 이벤트 리스너가 사라지는 문제가 발생할 수도 있습니다. createElement 방식이라면 이런 문제 없이 요소를 동적으로 추가하고, 이벤트도 안정적으로 관리할 수 있겠죠?!
물론, 템플릿 리터럴 방식도 정적인 UI를 렌더링할 때는 간결하고 효율적이지만, 보안과 유지보수를 고려하면 createElement 방식이 더 적절한 경우가 많다고 생각합니다!

> 저희는 '반복되는 로직'을 기준으로 컴포넌트를 나누었고, 먼저 Templates 파일을 기반으로 공통된 패턴을 정리한 후 컴포넌트로 분리했습니다. 
> 이 방식이 재사용성과 유지보수 측면에서 적절한지, 혹시 더 효과적인 분리 기준이 있을지 하리의 의견이 궁금합니다!

딱 봤을 때는 컴포넌트 분리의 기준을 "재사용성"에 포커스를 맞춰 분리하신 것 같아요. 그럼 재사용하지 못하는 컴포넌트는 분리하는 의미가 없을까요!?
유지보수 측면에서 적절한 컴포넌트는 어떻게 분리할 수 있을지도 고민해보시면 좋을 것 같습니당
저는 보통 컴포넌트를 UI 기준으로 분리하는 것 같아요! "하나의 컴포넌트에서는 하나의 역할을 수행하도록"

> 현재 E2E 테스트에서는 사용자의 주요 행동 시나리오를 기반으로 테스트를 작성했습니다. 현재 작성된 테스트 시나리오가 적절한 범위를 커버하고 있는지 하리의 의견이 궁금합니다! 혹시 놓치고 있는 중요한 시나리오가 있을까요?

굉장히 꼼꼼하게 작성해주셨던데요?! 개인적으로 의미 있는 E2E 테스트는 QA의 역할을 수행해야 한다고 생각하는데, 그런 의미에서 시나리오 기반으로 잘 작성해주신 것 같아요 :-)

> e2e 테스트 코드를 작성하던 중, cypress가 modal을 인식하지 못하는 문제로 인해, 다음과 같이 코드에 `{ force: true }` 속성을 부여했습니다. 
> 이렇게 요소가 안 보이더라도 강제 클릭하게 하는 기능을 사용하여, 테스트를 통과하는 것이 적절한 방식일지 고민입니다. 
> 하리의 의견은 어떠신가요?

음 사실 force 속성을 부여하는 건 다른 방법을 모두 시도해봐도 방법이 없을 때 시도해야 하는 마지막 방법이라고 생각합니당. 일단 e2e 테스트에서는 사용자 시나리오대로 테스트하는 게 필요한데, 모달이 display none으로 가려져 있어도 force 속성 때문에 테스트에서는 인식해버리면 존재 의의가 떨어지는 테스트가 되니까요!
근데 기린이 작성해주신 테스트 코드의 경우에는 모달을 인식하지 못한다기 보다는 viewport에 모달이 다 담기지 않아 생기는 문제인 것 같아요! [공식 문서](https://docs.cypress.io/api/commands/viewport) 쪽 참고해서 viewport를 키워서 한번 테스트해보셔도 좋을 것 같습니당 :-)

jeongyou commented:
안녕하세요, 하리! 방금 리뷰 요청 버튼을 실수로 눌러서 코멘트 남깁니다.😱 원래는 코멘트를 확인하고 리팩토링을 진행한 후에 리뷰 요청을 드리려고 했는데, 혹시 제가 바로 머지 하고 싶다는 뜻으로 오해하실까 봐 미리 말씀드립니다!

방금 눌렀던 리뷰 요청은 무시해 주시고, 수정 후 다시 요청드리겠습니다. 감사합니다!

jeongyou committed:
refactor: 프로젝트 폴더 구조 정리 및 역할별 파일 분리

- components/: 재사용 가능한 UI 컴포넌트를 관리
- constants/: 상수 값을 한 곳에서 관리
- data/: 레스토랑 데이터를 관리
- handlers/: 이벤트 핸들러를 모아서 관리
- store/: 상태 관리를 위한 store 파일을 별도로 관리
- styles/: 스타일 파일을 분리

jeongyou committed:
refactor: 폴더 구조 변경에 따른 import 경로 수정

jeongyou committed:
refactor: required 속성을 Boolean 값으로 변경

- required 속성을 문자열에서 Boolean 값으로 변경
- requiredClass 속성을 유지하여 CSS 적용 가능하도록 함

jeongyou committed:
test: DCI 패턴을 적용하여 Cypress 테스트 계층화 및 중복 코드 제거

- describe() 블록을 활용하여 '성공', '취소', '모달 동작', '유효성 검사' 테스트를 구분
- beforeEach()에서 중복된 모달 열기 코드 제거하여 테스트 가독성 향상
- 테스트 목적을 명확히 하기 위해 관련된 테스트들을 그룹화
- 터미널에서 실행 시 테스트 로그의 가독성을 개선

jeongyou committed:
refactor: script 태그 위치 조정하여 HTML 구조 개선

- `<script>` 태그를 `<div id='app'>` 내부에서 `<body>` 하단으로 이동
- HTML 구조를 명확하게 유지하고 코드 가독성 향상

jeongyou committed:
refactor: 모달 생성 방식을 template 태그 기반으로 변경

- 모달을 정적인 문자열 생성 방식에서 HTML의 <template> 태그를 활용하는 방식으로 리팩토링
- createModal() 함수가 template을 복사(cloneNode)하여 반환하도록 변경
- appendModal()에서 insertAdjacentHTML() 대신 appendChild() 사용하여 Node 객체 추가

jeongyou committed:
fix: label 태그의 for 속성을 올바르게 수정하여 입력 필드와 연결

- label 태그의 for 속성이 올바르게 설정되지 않아 select, input, textarea와 연결되지 않는 문제 수정
- 불필요하게 포함된 'text-caption' 문자열 제거
- 스타일 적용을 위해 class='text-caption'을 추가

jeongyou committed:
refactor: 이벤트 핸들러 구조 개선 및 main.js에 적용

- 기존 개별 이벤트 핸들러를 제거하고 개별 함수로 분리하여 재사용성 향상
- `registerEventHandlers()`를 도입하여 이벤트 핸들러 등록을 한 곳에서 관리하도록 변경
- `main.js`에서 기존 `addEventHandlers()` 호출을 제거하고 `eventHandlers.registerEventHandlers(addNewRestaurantItem)`를 호출하도록 수정
- 익명 함수 대신 명시적인 핸들러 함수를 사용하여 `removeEventListener()`를 통한 이벤트 해제 가능

jeongyou committed:
refactor: 이벤트 핸들러 구조 개선 및 main.js에 적용

- 기존 개별 이벤트 핸들러를 제거하고 개별 함수로 분리하여 재사용성 향상
- `registerEventHandlers()`를 도입하여 이벤트 핸들러 등록을 한 곳에서 관리하도록 변경
- `main.js`에서 기존 `addEventHandlers()` 호출을 제거하고 `eventHandlers.registerEventHandlers(addNewRestaurantItem)`를 호출하도록 수정
- 익명 함수 대신 명시적인 핸들러 함수를 사용하여 `removeEventListener()`를 통한 이벤트 해제 가능

jeongyou committed:
refactor: viewport 설정을 통해 불필요한 { force: true } 제거

jeongyou commented:
안녕하세요 하리! 기린입니다~. 꼼꼼한 리뷰 정말 감사합니다! 리뷰를 통해 기존 코드에서 미처 고려하지 못했던 부분들을 다시 한번 점검할 수 있었고, 더 나은 방향으로 개선할 수 있는 방법도 고민해 볼 수 있어서 많은 도움이 되었습니다. 좋은 피드백 주셔서 다시 한번 감사드립니다! 🙌

**각 리뷰 코멘트에 대한 수정 사항을 반영하고, 해당 댓글에 코멘트를 남겨놓았습니다!**

그리고 그 외에 따로 답변 주신 내용에 대해서도 리팩터링 해 보았습니다.
- 파일 폴더링 관련 리펙터링 -> [refactor: 프로젝트 폴더 구조 정리 및 역할별 파일 분리](https://github.com/woowacourse/javascript-lunch/pull/216/commits/847b4cadffd8fb17f07a63062021f60c2f3f7017)
- required를 문자열에서 Boolean으로 리펙터링 -> [refactor: required 속성을 Boolean 값으로 변경](https://github.com/woowacourse/javascript-lunch/pull/216/commits/50bc3416d979c5cde31f67969a602a49f039d599)
- viewport를 사용하여 fore:true 제거 -> [refactor: viewport 설정을 통해 불필요한 { force: true } 제거](https://github.com/woowacourse/javascript-lunch/pull/216/commits/b83155c8dd763cdb8a95f9bcaa6c76c5e77021ff)

아 그리고 템플릿 리터럴에서 createElement로의 변경은 현재 큰 문제가 되지는 않는다고 판단하여 시간이 되면 리팩터링 해보려고 합니다!

❓마지막으로 한 가지 궁금한 점이 있습니다.
```
it('모달 외부를 클릭하여 음식점 추가를 취소하는 시나리오 테스트', () => {
  cy.get('.modal-backdrop').click({ force: true });
  cy.get('.modal--open').should('not.exist');
});
```
- 이 부분에서 force: true를 여전히 사용하고 있습니다.
그런데 테스트할 때 z-index를 강제로 조정하면 force: true 없이도 정상적으로 동작하는 것 같습니다. 이런 경우에 z-index 조정을 통해 해결하는 것이 더 나을지, 아니면 그대로 force: true를 유지하는 것이 좋을지 하리의 의견이 궁금합니다!

끝입니다~

혹시 추가로 개선할 부분이 있다면 피드백 부탁드립니다. 🙌

LAH1203 commented:
> 이 부분에서 force: true를 여전히 사용하고 있습니다.
그런데 테스트할 때 z-index를 강제로 조정하면 force: true 없이도 정상적으로 동작하는 것 같습니다. 이런 경우에 z-index 조정을 통해 해결하는 것이 더 나을지, 아니면 그대로 force: true를 유지하는 것이 좋을지 하리의 의견이 궁금합니다!

force와 z-index 조정 모두 사용자가 직접 수행하는 시나리오가 아닌 시나리오를 통과하기 위해 테스트 환경 내 돔을 조작하는 동작이라, 최대한 자제하고 다른 방법을 강구해보시는 게 나을 것 같아요!

왜 `cy.get('.modal-backdrop').click()`으로 사용하셨을 때, `force: true`를 쓰라고 에러가 날까요? 이 원인은 찾아보셨을까요? z-index 쪽을 언급하신 걸 보면 대충은 알고 계실 것 같습니다~!
cypress는 내부적으로 click 실행 전에 몇 가지 검사를 수행합니다. [참고](https://docs.cypress.io/app/core-concepts/retry-ability#Implicit-Assertions)
- 사용자가 요소를 클릭할 수 있는지
- 요소가 보이는지
- 요소가 다른 요소 뒤에 있는지
- 요소가 disabled 속성을 가지고 있는지

지금 같은 경우는 modal-backdrop이 modal-container 뒤에 그려지기 때문이겠죠? 그럼 지금처럼 돔을 조작하지 않고 이 검사를 우회하는 방법은 없을까요?
한 가지 방법을 말씀드리면, 네이티브 클릭 메서드를 사용하는 방식입니다. cypress가 제공하는 클릭 메서드가 아니라 네이티브 요소가 가지고 있는 클릭 메서드를 사용하기 때문에 cypress가 click 호출 시 하는 검사를 별도로 수행하지 않습니다. 그리고 네이티브 요소를 받아오기 위해서는 [cypress의 then](https://docs.cypress.io/api/commands/then) 을 활용할 수 있을 것 같아요~! :)

LAH1203 reviewed (approved):
1단계 고생하셨습니다 기린 🦒

추가로 드린 코멘트들은 말씀해주신대로 리팩토링을 고민해보시면서!
목표를 이루려면 어떤 방법으로 사용하는 게 가장 좋을지, 각 방식에 대해 자세히 찾아보고 씹고 뜯고(?) 해보시면 좋을 것 같습니다 ㅎㅎㅎ

2단계도 화이팅!
